fbgemm_gpu/include/fbgemm_gpu/cpu_utils.h

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright (c) Intel Corporation.
+ * All rights reserved.
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+#pragma once
+
+#include <omp.h>
+#include <cstdint>
+#include <utility>
+
+template <typename T>
+using Radix_Sort_Pair = std::pair<T, T>;
+// histogram size per thread
+const int RDX_HIST_SIZE = 256;
+template <typename T>
+Radix_Sort_Pair<T>* radix_sort_parallel(
+    Radix_Sort_Pair<T>* inp_buf,
+    Radix_Sort_Pair<T>* tmp_buf,
+    int64_t elements_count,
+    int64_t max_value) {
+  int maxthreads = omp_get_max_threads();
+  alignas(64) int histogram[RDX_HIST_SIZE * maxthreads],
+      histogram_ps[RDX_HIST_SIZE * maxthreads + 1];
+  if (max_value == 0)
+    return inp_buf;
+  int num_bits = sizeof(T) * 8 - __builtin_clz(max_value);
+  unsigned int num_passes = (num_bits + 7) / 8;
+
+#pragma omp parallel
+  {
+    int tid = omp_get_thread_num();
+    int nthreads = omp_get_num_threads();
+
+    int* local_histogram = &histogram[RDX_HIST_SIZE * tid];
+    int* local_histogram_ps = &histogram_ps[RDX_HIST_SIZE * tid];
+    int elements_count_4 = elements_count / 4 * 4;
+    Radix_Sort_Pair<T>* input = inp_buf;
+    Radix_Sort_Pair<T>* output = tmp_buf;
+
+    for (unsigned int pass = 0; pass < num_passes; pass++) {
+      // Step 1: compute histogram
+      for (int i = 0; i < RDX_HIST_SIZE; i++)
+        local_histogram[i] = 0;
+
+#pragma omp for schedule(static)
+      for (int64_t i = 0; i < elements_count_4; i += 4) {
+        T val_1 = input[i].first;
```

Comment:
@ashokei have you tested if we have key and value as separate arrays? This may help vectorization. Although not entirely sure if vectorization will help this kind loop.